compiler/optimizer/OMRValuePropagation.hpp

```diff
@@ -354,12 +354,72 @@ class ValuePropagation : public TR::Optimization
+    * it is known to be the same as the component type of the array, and the class is
+    * not known to be extended; otherwise, the parameter is set to \c NULL.
+    * If the value is not \c NULL, any \ref TR::ArrayStoreCHK generated for this potential
+    * assignment could be NOPed based on whether the class is ever seen to be extended.
+    *
+    * The \c componentClassForCheck parameter is set to the component type of the array if
+    * the class of the object must be the same class as the component type of the array
+    * or a subtype of the component type; otherwise, the parameter is set to \c NULL.
+    * At run-time, a simple check that the actual component type of the array really is
+    * equal to \c componentClassForCheck can be used to avoid the need for a more extensive
+    * subtype check on any \ref TR::ArrayStoreCHK.
+    *
+    * At most one of \c storeClassForCheck and \c componentClassForCheck will be non-NULL.
+    * The value of \c storeClassForCheck can be set on a \ref TR::ArrayStoreCHK node
+    * via \ref TR::Node::setArrayStoreClassInNode(TR_OpaqueClassBlock*)
+    * The value of \c componentClassForCheck can be set on a \ref TR::ArrayStoreCHK node
+    * via \ref TR::Node::setArrayComponentClassInNode(TR_OpaqueClassBlock*)
+    *
+    * \param[in] arrayRef A \ref TR::Node representing an array reference, an element of
+    *                 which is being considered as the target of an assignment
+    * \param[in] arrayRef A \ref TR::Node representing an object reference that is being
+    *                 considered as the value to be assigned to the array element
+    * \param[out] mustFail An output parameter of type \c bool.  Set to \c true if the
+    *                 assignment under consideration would always result in an \c ArrayStoreException
+    * \param[out] storeClassForCheck An output parameter that is a pointer to \ref TR_OpaqueClassBlock
+    *                 representing what VP knows about the type of the object being stored
+    * \param[out] componentClassForCheck An output parameter that is a pointer to \ref TR_OpaqueClassBlock
+    *                 representing what VP knows about the type of the component type of the array
+    *
+    * \return \c false if Value Propagation determines that an \c ArrayStoreException would never
+    *                 be thrown by the assignment;\n
+    *         \c true otherwise
+    */
+   bool isArrayStoreCheckNeeded(TR::Node *arrayRef, TR::Node *objectRef, bool &mustFail,
+              TR_OpaqueClassBlock* &storeClassForCheck, TR_OpaqueClassBlock* &componentClassForCheck);
+
+   /**
+    * Determine the bounds and element size for an array constraint
+    *
+    * \param[in] arrayConstraint A \ref TR::VPConstraint for an array reference
+    * \param[out] lowerBoundLimit The lower bound on the size of the array
+    * \param[out] upperBoundLimit The upper bound on the size of the array
```

Comment:
Could the comments be elaborated to describe the values when they are not known? Such as for `lowerBoundLimit`, zero if not known; for `upperBoundLimit` max `Int32` if not known.

Reply:
Sounds good.  I'll do that.